### ISSUE TITLE
Potential buffer overflow in PyArray_NewFromDescr_int of ctors.c

### ISSUE THREAD

**Issue by Daybreak2019** [numpy/numpy#18939]
### Reproducing code example:
**Snippet**:
PyArray_NewFromDescr_int(...., **int nd**,......)
{
    ............... 
    if (descr->subarray) {
        PyObject *ret;
        npy_intp **newdims**[2*NPY_MAXDIMS];
        npy_intp *newstrides = NULL;
        **memcpy**(newdims, dims, nd*sizeof(npy_intp));
        if (strides) {
            newstrides = newdims + NPY_MAXDIMS;
            **memcpy**(newstrides, strides, nd*sizeof(npy_intp));
        }
        ........
}

### Error message:
When we run our analysis tool on NumPy,  a potential buffer-overflow problem was reported. See details below: 
**File**: numpy/core/src/multiarray/ctors.c
**Function**: PyArray_NewFromDescr_int
**Details** in [description](https://github.com/Daybreak2019/IssueReport/blob/master/Numpy/case-9532/9532.png)

Possible call path:
**1**. array_new -> PyArray_NewFromDescr_int
**2**. PyArray_Zeros -> PyArray_NewFromDescr_int
**3**. array_fromfile -> PyArray_FromFile -> PyArray_NewFromDescr_int

### NumPy/Python version information:
The main branch of NumPy

**Comment by eric-wieser** [comment]
I think this is probably safe at the call sites, but adding an assert to `PyArray_NewFromDescr_int` would probably be a good idea just to verify that

**Comment by eric-wieser** [comment]
Yes you're right, `PyArray_Zeros` is an API function, and does not perform any checking of `nd` before deferring to `PyArray_NewFromDescr_int`:

https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/core/src/multiarray/ctors.c#L2819-L2828

**Comment by seberg** [comment]
It might be a bit hard to abuse, but the check for `nd` should indeed be moved up here to be on the safe side.
